api/src/main/java/org/openmrs/api/context/Daemon.java

```diff
@@ -390,30 +360,114 @@ public void run() {
+		if (!ModuleFactory.isTokenValid(token)) {
+			throw new ContextAuthenticationException("Invalid token");
+		}
+		
+		return runInDaemonThreadInternal(callable);
+	}
+
+	/**
+	 * Executes the given runnable in a new thread that is authenticated as the daemon user.
+	 *
+	 * @param runnable an object implementing the {@link Runnable} interface to be run
+	 * @param token the token required to run code as the daemon user
+	 * @return the newly spawned {@link Thread}
+	 * @since 2.7.0
+	 */
+	@SuppressWarnings("squid:S1217")
+	public static Future<?> runInDaemonThreadWithoutResult(final Runnable runnable, DaemonToken token) {
+		if (!ModuleFactory.isTokenValid(token)) {
+			throw new ContextAuthenticationException("Invalid token");
+		}
+
+		return runInDaemonThreadInternal(runnable);
+	}
 	
 	/**
 	 * Executes the given runnable in a new thread that is authenticated as the daemon user and wait
 	 * for the thread to finish.
 	 *
 	 * @param runnable an object implementing the {@link Runnable} interface.
 	 * @param token the token required to run code as the daemon user
-	 * @since 1.9.2
+	 * @since 2.7.0
 	 */
 	public static void runInDaemonThreadAndWait(final Runnable runnable, DaemonToken token) {
-		Thread daemonThread = runInDaemonThread(runnable, token);
+		Future<?> daemonThread = runInDaemonThreadWithoutResult(runnable, token);
 		
 		try {
-			daemonThread.join();
+			daemonThread.get();
 		}
-		catch (InterruptedException e) {
-			//Ignore
+		catch (InterruptedException | ExecutionException e) {
+			// Ignored
 		}
 	}
 	
+	private static <T> Future<T> runInDaemonThreadInternal(Callable<T> callable) {
+		return OpenmrsThreadPoolHolder.threadExecutor.submit(() -> {
+			isDaemonThread.set(true);
+			try {
+				Context.openSession();
+				return callable.call();
+			}
+			finally {
+				try {
+					Context.closeSession();
+				} finally {
+					isDaemonThread.remove();
+				}
+			}
+		});
+	}
+	
+	private static Future<?> runInDaemonThreadInternal(Runnable runnable) {
+		// for Threads, we used to guarantee that Thread.start() was called before the function returned
+		// since we cannot guarantee that the executor actually started executing the thread, we use a CountDownLatch
+		// to emulate this behaviour when the user submits a Thread. Other runnables are unaffected.
+		CountDownLatch countDownLatch = getCountDownLatch(runnable instanceof Thread);
+
+		Future<?> result = OpenmrsThreadPoolHolder.threadExecutor.submit(() -> {
+			isDaemonThread.set(true);
+			try {
+				Context.openSession();
+				countDownLatch.countDown();
+				runnable.run();
+			}
+			finally {
+				try {
+					Context.closeSession();
+				} finally {
+					isDaemonThread.remove();
+				}
+			}
+		});
+		
+		try {
+			countDownLatch.await();
+		} catch (InterruptedException ignored) {}
+		
+		return result;
+	}
+
+	private static CountDownLatch getCountDownLatch(boolean isThread) {
+		return isThread ? new CountDownLatch(1) : new CountDownLatch(0);
```

Comment:
If we're using a Thread, we need to wait for the CountDownLatch, otherwise, we don't.